path_config.go

```diff
@@ -0,0 +1,232 @@
+// the responsibility of the caller to check if a config upgrade is required and,
+// if so, to upgrade the config
+func (b *backend) nonLockedOCIConfig(ctx context.Context, s logical.Storage) (*OCIConfigEntry, error) {
+	entry, err := s.Get(ctx, "config")
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result OCIConfigEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathConfigRead(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	configEntry, err := b.lockedOCIConfig(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if configEntry == nil {
+		return nil, nil
+	}
+
+	return &logical.Response{
+		Data: configEntry.ToResponseData(),
+	}, nil
+}
+
+// create a Config
+func (b *backend) pathConfigCreate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	homeTenancyId := data.Get(HomeTenancyIdConfigName).(string)
+	if strings.TrimSpace(homeTenancyId) == "" {
+		return logical.ErrorResponse("missing homeTenancyId"), nil
+	}
+
+	b.configMutex.Lock()
+	defer b.configMutex.Unlock()
+
+	configEntry, err := b.nonLockedOCIConfig(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	if configEntry != nil {
+		return logical.ErrorResponse("The specified config already exists"), nil
+	}
+
+	configEntry = &OCIConfigEntry{
+		HomeTenancyId: homeTenancyId,
+	}
+
+	if err := b.nonLockedSetOCIConfig(ctx, req.Storage, configEntry); err != nil {
+		return nil, err
+	}
+
+	var resp logical.Response
+
+	return &resp, nil
+}
+
+// update a Config
+func (b *backend) pathConfigUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	homeTenancyId := data.Get(HomeTenancyIdConfigName).(string)
+	if strings.TrimSpace(homeTenancyId) == "" {
+		return logical.ErrorResponse("missing homeTenancyId"), nil
+	}
+
+	b.configMutex.Lock()
+	defer b.configMutex.Unlock()
+
+	configEntry, err := b.nonLockedOCIConfig(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	if configEntry == nil {
+		return logical.ErrorResponse("The specified config does not exist"), nil
+	}
+
+	configEntry.HomeTenancyId = homeTenancyId
+
+	if err := b.nonLockedSetOCIConfig(ctx, req.Storage, configEntry); err != nil {
+		return nil, err
+	}
+
+	var resp logical.Response
+	return &resp, nil
+}
+
+// delete a Config
+func (b *backend) pathConfigDelete(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	b.configMutex.Lock()
+	defer b.configMutex.Unlock()
+
+	return nil, req.Storage.Delete(ctx, "config")
+}
+
+// Struct to hold the information associated with an OCI config
+type OCIConfigEntry struct {
+	HomeTenancyId string `json:"homeTenancyId" `
+}
+
+func (r *OCIConfigEntry) ToResponseData() map[string]interface{} {
+	responseData := map[string]interface{}{
```

Comment:
Just embed this is the response and remove the helper function.

Reply:
Done